### DISCUSSION TITLE
IJob not triggered when something is injected to it

### DISCUSSION THREAD

**Discussion by vlapenkov** [quartznet/quartznet#977] [General] [answered]
**When anything is injected in IJob implementation class, it.is not called**

Asp.net core 3.1 and 
```
 <PackageReference Include="Quartz.AspNetCore" Version="3.1.0" />
<PackageReference Include="Quartz.Extensions.DependencyInjection" Version="3.1.0" />
```

Event triggers every 10 seconds when nothing is injected to ExampleJob, but not triggers when logger is injected.

Startup.cs
```
public void ConfigureServices(IServiceCollection services)
        {            services.AddTransient<ExampleJob>();                      
            services.AddQuartz(async( q) =>
            {
                // get a scheduler
                IScheduler scheduler = await StdSchedulerFactory.GetDefaultScheduler();
                await scheduler.Start();

                IJobDetail job = JobBuilder.Create<ExampleJob>().Build();     

                ITrigger trigger = TriggerBuilder.Create()
                .WithIdentity("myTrigger", "group1")
                .StartNow()
                .WithSimpleSchedule(x => x.WithIntervalInSeconds(10).RepeatForever()).Build();

                await scheduler.ScheduleJob(job, trigger);
            }            
            );
            services.AddControllers();
}

```
ExampleJob.cs
```
public class ExampleJob : IJob
    {
        private ILogger<ExampleJob> _logger;

        public ExampleJob(ILogger<ExampleJob> logger)
        {
            _logger = logger;
        }
        public async Task Execute(IJobExecutionContext context)
        {
            Console.WriteLine($"Logger {DateTime.Now}");
       
           _logger.LogInformation($"Logger {DateTime.Now}");
        }
    }
```

**Comment by lahma** [comment, marked as answer]
Please follow [the documentation](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#di-aware-job-factories) and add the job via `q.AddJob...`. This will register it to DI. You should not take reference to scheduler inside the configuration while you are still building the configuration. It's especially bad now when the lambda is `async void`, it won't be even `await`ed.

**Reply by vlapenkov** [reply]
When should i use AddQuartzServer?
    services.AddQuartzServer(options =>
    {
        // when shutting down we want jobs to complete gracefully
        options.WaitForJobsToComplete = true;
    });

**Reply by lahma** [reply]
If you want Quartz to register a [background service](https://github.com/quartznet/quartznet/blob/master/src/Quartz.Extensions.Hosting/QuartzHostedService.cs) that runs the scheduler, basically it automates starting and stopping of the scheduler.
